packages/plugins/typescript/operations/tests/ts-documents.spec.ts

```diff
@@ -7423,5 +7423,108 @@ function test(q: GetEntityBrandDataQuery): void {
         export type UserFragmentFragment = { __typename?: 'User', id: string } & { ' $fragmentName'?: 'UserFragmentFragment' };
       `);
     });
+
+    it("'mask' with @unmask yields correct types", async () => {
+      const ast = parse(/* GraphQL */ `
+        query {
+          me {
+            ...UserFragment @unmask
```

Comment:
Apollo Client will have support for `@unmask(mode: "migrate")` which will generate runtime warnings if you access a field that would otherwise be masked. I'd love for this to generate the field and mark it with `@deprecated` so that users get a strikethrough in their editor.

I struggled to add this however as it felt the change needed would be a fairly large change to accomodate the data needed to add this. I'd love to discuss viable ways to handle this, both from an implementation standpoint, and in a way that makes sense for GraphQL Codegen. 

I realize this is quite specific to Apollo Client so I'd be happy to discuss viable options for adding this functionality without shifting it too much toward Apollo Client specifically (and FYI, we are open to changing the name of the argument on our end if it makes sense to do so).

Reply:
I think `@deprecated` is a nice idea. How much of a change would this be?

IIRC GraphQL Codegen's fragment masking feature is just faux type, and the props are still available at runtime, unlike Apollo Client's feature. So, `@unmask` _could_ make sense if the user tries to widen the type. Maybe the `mode` could be more generic like `'report-invalid-usage'` or something similar?

Reply:
> I think @deprecated is a nice idea. How much of a change would this be?

The trouble here is that I need to be able to traverse the selection set of the named fragment and add `/** @deprecated */` above the generated properties for all fields in that type, but there is no real way to hold onto that "state" so that the code further down in the plugin can add this correctly. It felt like I'd need a large change in this plugin to be able to track that somehow so that by the time the string is created, we know whether to add that comment or not. Its just that seems to happen much further down in this plugin IIRC. If you have any suggestions on how best to approach this, that would be much appreciated.

That said, this is purely a DX thing. I care more that the fields are available in the type when `@unmask` is used, which is already working. I just thought it would be a nice touch to also show a strikethrough in the editor by marking the field as deprecated so that we have a nice visual to which fields will be removed once you remove the `@unmask(mode: "migrate")` directive from the named fragment.

> IIRC GraphQL Codegen's fragment masking feature is just faux type, and the props are still available at runtime, unlike Apollo Client's feature. So, `@unmask` could make sense if the user tries to widen the type.

I believe you're right. If I remember looking at the generated `useFragment` function from codegen's fragment masking feature, its just an identity function. Apollo Client on the other hand will either hide the properties in the returned type, or make them available if it detects an `@unmask` directive at a particular named fragment. I'd love for the types to reflect that as well, hence this change.

With the suggestion to add the additional config option though, I'm less concerned that this feature will leak outside of Apollo's use case. If down the line you adopt this directive as part of the fragment masking feature, then great! But I think using that config option first to introduce this behavior makes most sense 🙂 

> Maybe the mode could be more generic like 'report-invalid-access' or something similar?

We called that argument `mode: "migrate"` to signal to the user that using is used for migration purposes. The feature, when used with that mode in Apollo Client, will also console warn at runtime. I'd like to keep the name unless we have a good reason to change it since there is that runtime behavior. We will have a ton of users with large apps that can't use data masking immediately, so we'll have a codemod that will go apply `@unmask(mode: "migrate")` to all named fragments app-wide so that users can start the process of migrating to masking without having to do it at once. We figured it would be nice to support that whole use case in the TypeScript story as well.

Again, since we will be going with the new config option, I think the argument name is less important now that it will be marked as an apollo feature.

Reply:
> Again, since we will be going with the new config option, I think the argument name is less important now that it will be marked as an apollo feature.

Agreed 👍  Sorry, I left the `report-invalid-access` comment before chatting to the team about using a specific `apolloUnmask` config and was thinking "if we could use it for our useFragment, what may a good name be?"

Reply:
I'm tabling the `@unmask(mode: "migrate")` with the `@deprecated` output for now. I'd love to add this in a future improvement, but having some form of `@unmask` support is the biggest thing for me right now. Would love to revisit when I have more time to learn the codegen codebase!